toolchain/check/name_component.h

```diff
@@ -36,12 +36,18 @@ struct NameComponent {
   SemIR::InstBlockId params_id;
   SemIR::InstBlockId param_patterns_id;
 
+  // The return slot.
+  SemIR::InstId return_slot_pattern_id;
+  SemIR::InstId return_slot_id;
```

Comment:
It seems a bit strange to include this in `NameComponent` given that it never appears in a name qualifier. I would have expected the return type to be a separate part of function syntax that appears after the name, rather than being treated as part of the name itself. Is it important that we handle this here rather than only in logic that's specific to function declarations? If so, have you thought about how we would model `fn F() => expr;`?

Reply:
> A name qualifier never has a return type, but IIUC it never has implicit parameters either, and yet `NameComponent` includes implicit parameters.

A `NameComponent` can include implicit parameters when used as a qualifier:

```carbon
class C[T:! type](N:! T) {
  fn F();
}
fn C[T:! type](N:! T).F() {}
```

> It may not be important now, but in the future I think it's likely that we'll want to have a type that can represent any full-pattern, whether or not it's prefixed with a name. Right now that's easy: it's just a matter of factoring out a struct consisting of everything below name_id (or maybe below last_param_node_id). If the return slot members are in a different struct, that becomes difficult if not impossible.

I see. For right now, it seems that the place where this comes up is `PopNameComponent`, which is calling `CalleePatternMatch` itself and handling the return pattern and return slot parameter in addition to the declared parameters. Doing that all together seems logical given that it's effectively computing the calling convention of the function (in the case where the name is for a function). I wonder whether we could factor that process a bit differently so we only deal with return types along a function-specific codepath (eg, maybe add a `PopFunctionSignature` that handles the return type, and returns a `FunctionSignature` containing a `NameComponent` and a return type).

But in any case: while this representation seems to me to be conflating two similar but different things, I'm OK with us using it for now and revisiting this once the broader pattern matching support you mentioned is built. Please add a TODO to reconsider including the return information in `NameComponent`.

Reply:
OK, I've added a TODO.